backend/ipp.c

```diff
@@ -2751,7 +2751,10 @@ new_request(
 					/* I - Printer supports print-color-mode */
 {
   ipp_t		*request;		/* Request data */
+  const char	*color_attr_name = print_color_mode_sup ? "print-color-mode" : "output-mode";	/* Name for color IPP attribute */
```

Comment:
This should be in _cupsConvertOptions in cups/ppd-cache.c

Reply:
@michaelrsweet however _cupsConvertOptions() are not called for remote CUPS queues, which is a problem in case the remote destination is a non-driverless queue. IMHO we need to consolidate sent options, in case an application does not do it, otherwise the output will be set by the last color related attribute...

Other place which can work for this case could ```create_job()``` in scheduler, but handling this in IPP backend looked more appropriate, because we can handle this on place for all use cases via IPP (temp queue, permanent queue to printer and permanent queue to server)

So should I implement this for ```create_job()```? Or update _cupsConvertOptions() to don't use PPD cache, so it could be useful for remote queue as well? Or something else you would propose?

Reply:
@michaelrsweet the user confirmed the PR works for Okular as well as it is now - https://bugzilla.redhat.com/show_bug.cgi?id=1985917 - moving the code back ```_cupsConvertOptions()``` breaks the color printing again.

Reply:
I agree that the code suits better into _cupsConvertOptions(), but we have to ensure consistent options values for all destinations (in case applications are not up to it...) alongside consistent interpretation of PPD and IPP options - moving the code from _cupsConvertOptions() does the both in one swoop.

Would it be okay to move the code into a new function - f.e. _cupsValidateValues() - which would correct the values based on preferences and this options would be convert to IPP attributes in _cupsConvertOptions() and cupsEncodeOptions2()?

Reply:
I wouldn't want to mess with cupsEncodeOptions since that could have some nasty side-effects.

_cupsConvertOptions also gets used on macOS to (separately) do a Validate-Job request for transaction-based printing - I know they aren't exactly keeping up with us anymore but I don't want to create unnecessary churn. The goal of _cupsConvertOptions is to convert PPD options to IPP standard attributes and values.

It sounds like what we need here (for the CUPS to CUPS case) is a _cupsConvertAttributes function that converts IPP attributes and values to PPD options and choices. This can be applied in the CUPS-to-CUPS remote printing case, and you can probably refactor cupsMarkOptions to use it internally to handle all of the standard IPP attributes and values.

Reply:
> I wouldn't want to mess with cupsEncodeOptions since that could have some nasty side-effects.
> 

@michaelrsweet  Ack.

> _cupsConvertOptions also gets used on macOS to (separately) do a Validate-Job request for transaction-based printing - I know they aren't exactly keeping up with us anymore but I don't want to create unnecessary churn. The goal of _cupsConvertOptions is to convert PPD options to IPP standard attributes and values.
> 

Ok.

> It sounds like what we need here (for the CUPS to CUPS case) is a _cupsConvertAttributes function that converts IPP attributes and values to PPD options and choices. This can be applied in the CUPS-to-CUPS remote printing case, and you can probably refactor cupsMarkOptions to use it internally to handle all of the standard IPP attributes and values.

IMO we don't need to convert attributes to PPD options - we just have to make sure the related PPD options/IPP attributes have the matching values before we pass them further. The different values for related options and attributes causes inconsistencies in the printing in the end, making it more difficult to predict what value will be applied.

My core idea behind CUPS to CUPS options+attributes transfer is to keep the changes at minimum (we don't know which options/attributes the other side will need in the end) and only validate the related options+attributes values.

So I propose the following:

- revert the current changes back to ```_cupsEncodeOptions()```
- implement ```_cupsValidateOptionValues()``` which will be used for CUPS-to-CUPS transfer and will adjust IPP attributes values, which are taken from defaults, to values from PPD options sent by an application - the result will be passed to ```cupsEncodeOptions2()``` as it is now.

WDYT?

Reply:
@zdohnal That could work, I'll take a look at the new changes when you have them...